mllib/src/main/scala/org/apache/spark/ml/classification/DecisionTreeClassifier.scala

```diff
@@ -0,0 +1,155 @@
+
+  override def setMinInstancesPerNode(minInstancesPerNode: Int): DecisionTreeClassifier =
+    super.setMinInstancesPerNode(minInstancesPerNode)
+
+  override def setMinInfoGain(minInfoGain: Double): DecisionTreeClassifier =
+    super.setMinInfoGain(minInfoGain)
+
+  override def setMaxMemoryInMB(maxMemoryInMB: Int): DecisionTreeClassifier =
+    super.setMaxMemoryInMB(maxMemoryInMB)
+
+  override def setCacheNodeIds(cacheNodeIds: Boolean): DecisionTreeClassifier =
+    super.setCacheNodeIds(cacheNodeIds)
+
+  override def setCheckpointInterval(checkpointInterval: Int): DecisionTreeClassifier =
+    super.setCheckpointInterval(checkpointInterval)
+
+  override def setImpurity(impurity: String): DecisionTreeClassifier = super.setImpurity(impurity)
+
+  override protected def train(
+      dataset: DataFrame,
+      paramMap: ParamMap): DecisionTreeClassificationModel = {
+    val categoricalFeatures: Map[Int, Int] =
+      MetadataUtils.getCategoricalFeatures(dataset.schema(paramMap(featuresCol)))
+    val numClasses: Int = MetadataUtils.getNumClasses(dataset.schema(paramMap(labelCol))) match {
+      case Some(n: Int) => n
+      case None => throw new IllegalArgumentException("DecisionTreeClassifier was given input" +
+        s" with invalid label column, without the number of classes specified.")
+        // TODO: Automatically index labels.
+    }
+    val oldDataset: RDD[LabeledPoint] = extractLabeledPoints(dataset, paramMap)
+    val strategy = getOldStrategy(categoricalFeatures, numClasses)
+    val oldModel = OldDecisionTree.train(oldDataset, strategy)
+    DecisionTreeClassificationModel.fromOld(oldModel, this, paramMap)
+  }
+
+  /** Create a Strategy instance to use with the old API. */
+  override private[ml] def getOldStrategy(
+      categoricalFeatures: Map[Int, Int],
+      numClasses: Int): OldStrategy = {
+    val strategy = super.getOldStrategy(categoricalFeatures, numClasses)
+    strategy.algo = OldAlgo.Classification
+    strategy.setImpurity(getOldImpurity)
+    strategy
+  }
+}
+
+object DecisionTreeClassifier {
+  /** Accessor for supported impurities */
+  final val supportedImpurities: Array[String] = TreeClassifierParams.supportedImpurities
+}
+
+/**
+ * :: AlphaComponent ::
+ *
+ * [[http://en.wikipedia.org/wiki/Decision_tree_learning Decision tree]] model for classification.
+ * It supports both binary and multiclass labels, as well as both continuous and categorical
+ * features.
+ */
+@AlphaComponent
+class DecisionTreeClassificationModel private[ml] (
+    override val parent: DecisionTreeClassifier,
+    override val fittingParamMap: ParamMap,
+    override val rootNode: Node)
+  extends PredictionModel[Vector, DecisionTreeClassificationModel]
+  with DecisionTreeModel with Serializable {
+
+  require(rootNode != null,
+    "DecisionTreeClassificationModel given null rootNode, but it requires a non-null rootNode.")
+
+  override protected def predict(features: Vector): Double = {
```

Comment:
Let's mark `DecisionTreeClassificationModel` final. Otherwise, if we make `predict` protected now, changing it to public will be a break change in the future.